---
title: "SHINYREPS_PROJECT"
output:
  html_document:
    toc: true
    toc_float: true
    css: styles.css
---

<div class="contentbox">


# Description




```{r setup, echo=F, result='hide', error=F, warning=F, message=F}

# source helper functions
source("mps.shinyrep.helpers.R")

# load required packages
notYetAttached <- attach_packages(c("rmarkdown", "knitr", "Cairo", "gplots", "ggplot2", "plotly", "RColorBrewer", "ggrepel", "parallel",
                  "grid", "gridExtra", "rtracklayer", "ggbeeswarm", "dplyr", "tidyr", "forcats", "gtools", "Biobase",
                  "scales", "kableExtra", "reshape2", "vipor", "corrplot", "limma", "viridis", "pkgmaker", "DT", "htmltools",
                  "igraph", "shinydashboard"))

# load global variables
loadGlobalVars(f="shinyReports.txt")


# set options
options(stringsAsFactors=FALSE)
CORES <- 2
pal   <- brewer.pal(9, "Set1")
pal_rb <- colorRampPalette(c(pal[1], "white", pal[2]))(20)
pal_y  <- colorRampPalette(c("black", "yellow"))(100)

knitr::opts_chunk$set(cache=F, echo=F, warning=F, message=F, dev='CairoPNG')

theme_set(theme_bw() + theme(axis.text=element_text(colour="grey30",size=12),
									axis.title=element_text(colour="grey30",size=14),
									plot.title=element_text(size=14,hjust=0.5),
									plot.subtitle=element_text(size=12,hjust=0.5),
									legend.text=element_text(size=12,colour="grey30"),
									legend.title=element_text(size=12,colour="grey30")))


run_cutadapt <- dir.exists(SHINYREPS_CUTADAPT_STATS)
run_2ndQC_after_cutadapt <- F
run_pear_assembly <- dir.exists(SHINYREPS_PEAR_LOG)
run_whitelist_extraction <- dir.exists(SHINYREPS_UMIEXTRACT_LOGWL)
run_nested_extraction <-any(grepl("nested", list.files(SHINYREPS_UMIEXTRACT_LOG))) 
PSIavailable <- file.exists(file.path(SHINYREPS_MPSPROFILING_OUT, "MPSprofiling.RData")) # PSI calculation applied?
  
## load targets file  
targets <- read.delim(SHINYREPS_TARGET, sep="\t")
  
# print targets file
DT::datatable(targets, caption="targets file (sample sheet) used for analysis")

if(PSIavailable) {
  # load PSI tables if available
  load(file.path(SHINYREPS_MPSPROFILING_OUT, "MPSprofiling.RData"))

  # format numbers in data tables and insert new lines in header
  for(i in names(bynuc_all)) {
  numcols <- sapply(bynuc_all[[i]], is.numeric)
  bynuc_all[[i]][,numcols] <- lapply(bynuc_all[[i]][,numcols], signif, 5)
  colnames(bynuc_all[[i]]) <- gsub("_", "\n", colnames(bynuc_all[[i]]))
  numcols <- sapply(byaa_all[[i]], is.numeric)
  byaa_all[[i]][,numcols] <- lapply(byaa_all[[i]][,numcols], signif, 5)
  colnames(byaa_all[[i]]) <- gsub("_", "\n", colnames(byaa_all[[i]]))
  }
}

```

# Quality control

## FastQC of all reads 

The raw sequence reads of all samples are analysed with the popular FastQC tool (http://www.bioinformatics.babraham.ac.uk/projects/fastqc/).

1. The "Read qualities" Box-Whisker plots show the range of quality values across all base positions:
    (i) The central red line is the median value,
    (ii) The yellow box represents the inter-quartile range (25-75%),
    (iii) The upper and lower whiskers represent the 10% and 90% points,
    (iv) The blue line represents the mean quality.
The y-axis on the graph shows the Phred quality scores, which are logarithmically related to the base-calling error probabilities. The higher the score the better the base call. The background of the graph divides the y axis into very good quality calls (green), calls of reasonable quality (orange), and calls of poor quality (red). Typically, the majority of calls on all base positions fall into the green area.

2. The "Sequence bias" plots show the proportion of each base (% G, A, T and C) at each position. In a random library there would be little difference between the positions of a sequence run, so the lines in this plot should run parallel with each other. But most RNA-seq libraries show sequence imbalance in the first 10-12 read positions due to RT priming biases, which should however look fairly similar in all samples.

3. The "GC Content" plots show the GC% distribution of all reads (red lines) and the ideal normal distribution based on the data (blue lines). Typically, the red and blue lines tightly overlap and look essentially the same in all samples. An unusually shaped distribution could indicate a contaminated library.

```{r FastQC_paragraph1, echo=F, results='asis', error=F, warning=F, message=F}
##### parameters to set:
# Select samples for which you would like to include fastqc results in the report. For huge experiments, 
# you may want to restrict the total number of plots. If you provide a regular expression in 'samplePattern' only those 
# filenames will be included which match this expression, e.g. setting samplePattern="R1" yields only those fastq 
# files containing read1 of a read pair. 
# For samplePattern=NULL all fastq files will be included.
# If you want to exclude samples according to the given certain pattern (instead of including), use parameter excludePattern. 
# If you set argument 'maxno', the maximum sample number will be restricted accordingly to the first 'maxno' plots. 
##### 

cat(MPShelper.Fastqc(web=F, samplePattern=SHINYREPS_SAMPLEPATTERN1, maxno=SHINYREPS_MAXNO), sep="\n")
```


```{r fastqscreen_paragraph, echo=F, results='asis', fig.width=10, error=F, warning=F, message=F}
cat("## Competetive mapping to rRNAs and other contaminants

A competitive mapping to the reference genome and known rRNAs (H. sapiens, M. musculus, C. elegans, D. melanogaster, D. rerio, X. tropicalis) as well as frequently occurring Mycoplasma species (M. arginini, M. fermentans, M. hominis, M. hyorhinis, M. orale, and Acholeplasma laidlawii), PhiX and ERCC spike-in sequences, E. coli and B. taurus (bovine) was performed with FastQScreen (http://www.bioinformatics.babraham.ac.uk/projects/fastq_screen/).

In the following plot only relevant contaminants are shown. A contaminant is considered relevant if it consumes `r SHINYREPS_FASTQSCREEN_PERC`% of reads or more in at least one sample. The 'one genome' label refers to the amount of reads mapping to that genome only, no matter if they are mapping uniquely or multiple times. The 'multiple genome' label refers to the amount of read mapping to this genome but also to other genomes in the screening regardless of whether they are uniquely or multi mapping.\n\n")

fastqscreen_plots <- MPShelper.fastqscreen(perc.to.plot=SHINYREPS_FASTQSCREEN_PERC, ncol=2)
```

```{r fastqscreen_paragraph_plot, echo=F,results='asis',error=F,warning=F,message=F,fig.align='center', fig.height=min(170, (round((1.8*fastqscreen_plots$no.of.rows))+1)*fastqscreen_plots$no.of.samples/8+1)}
  if(is.null(fastqscreen_plots$errortext)) {
    plot(fastqscreen_plots$p.category.wrap)
    } else {
      cat(fastqscreen_plots$errortext)
      }
```




```{r cutadapt, echo=F, results='asis', error=F, warning=F, message=F, eval=run_cutadapt}
##### parameters to set:
# define categories of targets.txt to be used for dot color in the plot (one plot per element of colorByFactor will be created). 
# The function will try to map the data from targets$sample to the cutadapt log file names (for this, the unique part given in  
# sampleColumnName must be a substring of file the name). Otherwise pruned file names will be used by default. 
# In case of pooled library design the targets object will be modified accordingly. 
#####

cat("## Adapter trimming with Cutadapt

The following plot shows the amount of reads trimmed for the selected adapter sequences. The column 'tooshort' gives the percentage of reads removed due to a length cutoff after trimming. Additional to the column 'trimmed', which is for all adapters combined, there are columns for every adapter giving the percentage of reads trimmed for this adapter.\n")

MPShelper.cutadapt(targetsdf=targets, colorByFactor=NULL)
```



```{r FastQC_paragraph2, echo=F, results='asis', error=F, warning=F, message=F, eval=run_cutadapt & run_2ndQC_after_cutadapt}
##### parameters to set:
# Select samples for which you would like to include fastqc results in the report. For huge experiments, 
# you may want to restrict the total number of plots. If you provide a regular expression in 'samplePattern' only those 
# filenames will be included which match this expression, e.g. setting samplePattern="R1" yields only those fastq 
# files containing read1 of a read pair. 
# For samplePattern=NULL all fastq files will be included.
# If you want to exclude samples according to the given certain pattern (instead of including), use parameter excludePattern. 
# If you set argument 'maxno', the maximum sample number will be restricted accordingly to the first 'maxno' plots. 
##### 

cat("### FastQC after adapter trimming\n")

cat(MPShelper.Fastqc(web=F, subdir="trimmed", samplePattern=SHINYREPS_SAMPLEPATTERN1, maxno=SHINYREPS_MAXNO), sep="\n")
```



```{r pear, echo=F, results='asis', error=F, warning=F, eval=run_pear_assembly}
##### parameters to set:
# include only files generated with umi_tools whitelist
samplePattern=NULL 
# define categories of targets.txt to be used for dot color in the plot (one plot per element of colorByFactor will be created). 
colorByFactor = SHINYREPS_COLORBYFACTOR 
#####

cat("# Read assembly with PEAR\n")
cat("\n\nThe 2 reads of a readpair are assembled to a single read. PEAR stats are given below.\n\n")

MPShelper.pear(samplePattern=samplePattern, colorByFactor=colorByFactor, targetsdf=targets)
```



```{r umitools_whitelist, echo=F, results='asis', error=F, warning=F, eval=run_whitelist_extraction}
##### parameters to set:
# include only files generated with umi_tools whitelist
samplePattern="whitelist" 
# define categories of targets.txt to be used for dot color in the plot (one plot per element of colorByFactor will be created). 
colorByFactor = SHINYREPS_COLORBYFACTOR 
#####

cat("# Whitelist extraction with UMI-tools\n")
cat("\nThe whitelist option offers to estimate likely true cell barcodes which can then be used to filter (and optionally correct) the barcodes extracted from the fastq files. The obtained whitelist contains the observed deviations from the 'real' variable regions (including their frequency). The extracted whitelists for each sample are available as tsv files in the", gsub(SHINYREPS_PROJECT, "", SHINYREPS_UMIEXTRACT_LOGWL),"folder. They contain 4 tab-separated columns: 1) whitelisted cell barcode, 2) Other cell barcode(s) (comma-separated) to correct to the whitelisted barcode, 3) Count for whitelisted cell barcodes, 4) Count(s) for the other cell barcode(s) (comma-separated).\n")

cat("\n\nKnee plots for barcode selection:\n")
MPShelper.whitelistextraction(samplePattern=samplePattern)

cat("\n\n\nUMI-tools stats for whitelist extraction:\n")
MPShelper.umiextract(samplePattern=samplePattern, colorByFactor=colorByFactor, targetsdf=targets)

```



# Extract barcodes with UMI-tools

Barcodes and UMIs (if applicable) are extracted from sequences in fastq files and are appended to the corresponding read names for later counting. UMIs will be used for deduplication of barcodes. Mismatches are allowed in UMIs as defined in project settings. 

```{r umiextract, echo=F, results='asis', error=F, warning=F}
##### parameters to set:
# exclude files from optional nested run
samplePattern=SHINYREPS_PATTERNUMITOOLS 
# define categories of targets.txt to be used for dot color in the plot (one plot per element of colorByFactor will be created). 
colorByFactor = SHINYREPS_COLORBYFACTOR 
#####

MPShelper.umiextract(excludePattern=samplePattern, colorByFactor=colorByFactor, targetsdf=targets)
```


```{r umiextract_nested, echo=F, results='asis', error=F, warning=F, eval=run_nested_extraction}
##### parameters to set:
# include files from optional nested run only
samplePattern=SHINYREPS_PATTERNUMITOOLS 
# define categories of targets.txt to be used for dot color in the plot (one plot per element of colorByFactor will be created). 
colorByFactor = SHINYREPS_COLORBYFACTOR 
#####

cat("## Barcode extraction for 2nd pattern\n")
cat("Barcode extraction statistics for 2nd run of umi_tools extract.\n")

MPShelper.umiextract(samplePattern=samplePattern, colorByFactor=colorByFactor, targetsdf=targets)
```




```{r countdata, results='asis', eval=PSIavailable}
cat("# Raw count data\n")
cat("## Histograms of count data\n")

histofiles <- list.files(file.path(SHINYREPS_MPSPROFILING_OUT, "plots"), full.names =T)
histofiles <- histofiles[grep("histogram", basename(histofiles))] # filter for histograms
histofiles <- histofiles[grep("_raw", basename(histofiles))] # filter for raw count files

for(f in histofiles) {
 # include plots in report    
  cat(paste0("![plot of ", basename(f), "](", f, ")"))
}
```


```{r PSIperNuc, results='asis', eval=PSIavailable}
cat("## PSIs calculated per nucleotides
    
Protein stability indices have been calculated per nucleotide sequence from normalized count data. The statistics given in the table are calculated from count data of all fractions per sub_experiment and sequence. 'nfractions' gives the number of fractions with counts detected, while 'totalfractions' contains the total number of fractions for this sub_experiment.\n")

 htmltools::tagList(
   lapply(unique(targets$experiment), function(x) {
      dplyr::filter(bynuc_all[["raw"]], experiment==x) %>%
                      head(n=as.numeric(SHINYREPS_MAX_TABLE_ROWS)) %>%
                      DT::datatable(caption=paste("PSI per nucleotide for experiment", x, 
                                                  if(as.numeric(SHINYREPS_MAX_TABLE_ROWS)<nrow(filter(bynuc_all[["raw"]], experiment==x)))
                                                  {paste("(max", as.numeric(SHINYREPS_MAX_TABLE_ROWS), "entries)")} else{""})
                                                    )
      })
    )
  
```


```{r PSIperAA, results='asis', eval=PSIavailable}
cat("## PSIs calculated per amino acids
    
Median protein stability indices per amino acid sequence have been calculated as median of PSIs from all nucleotide sequences translated into the same amino acid sequence. Additionally, pooled PSIs are calculated directly from normalized count data by pooling counts from all nucleotide sequences translated into the same amino acid sequence. The statistics given in the table refer to the pooled approach. 'nsequences' gives the number of nucleotide sequences tranlated into the same amino acid sequence, 'nfractions' gives the number of fractions with counts detected, while 'totalfractions' contains the total number of fractions for this sub_experiment.\n")

 htmltools::tagList(
   lapply(unique(targets$experiment), function(x) {
      dplyr::filter(byaa_all[["raw"]], experiment==x) %>%
                      head(n=as.numeric(SHINYREPS_MAX_TABLE_ROWS)) %>%
                      DT::datatable(caption=paste("PSI per amino acid for experiment", x, 
                                                  if(as.numeric(SHINYREPS_MAX_TABLE_ROWS)<nrow(filter(byaa_all[["raw"]], experiment==x)))
                                                  {paste("(max", as.numeric(SHINYREPS_MAX_TABLE_ROWS), "entries)")} else{""})
                                                    )
      })
    )
```



```{r countdata_bgsubt, results='asis', eval=(PSIavailable && "bgsubt" %in% names(counts_all))}
cat("# Background subtracted count data\n")
    
cat("## Histograms of count data after background subtraction\n

To remove background noise, a Gaussian mixture model is fitted to each sample and then the mean of the background distribution subtracted from the raw count data. All downstream processing steps are performed as for the raw count data.\n")

histofiles <- list.files(file.path(SHINYREPS_MPSPROFILING_OUT, "plots"), full.names =T)
histofiles <- histofiles[grep("histogram", basename(histofiles))] # filter for histograms
histofiles <- histofiles[grep("_bgsubt", basename(histofiles))] # filter for bgsubt count files

for(f in histofiles) {
 # include plots in report    
  cat(paste0("![plot of ", basename(f), "](", f, ")"))
}
```


```{r PSIperNuc_bgsubt, results='asis', eval=(PSIavailable && "bgsubt" %in% names(bynuc_all))}
cat("## PSIs calculated per nucleotides
    
Protein stability indices have been calculated per nucleotide sequence from normalized count data after background subtraction.\n")

 htmltools::tagList(
   lapply(unique(targets$experiment), function(x) {
      dplyr::filter(bynuc_all[["bgsubt"]], experiment==x) %>%
                      head(n=as.numeric(SHINYREPS_MAX_TABLE_ROWS)) %>%
                      DT::datatable(caption=paste("PSI per nucleotide for experiment", x, 
                                                  if(as.numeric(SHINYREPS_MAX_TABLE_ROWS)<nrow(filter(bynuc_all[["bgsubt"]], experiment==x)))
                                                  {paste("(max", as.numeric(SHINYREPS_MAX_TABLE_ROWS), "entries)")} else{""})
                                                    )
      })
    )
```


```{r PSIperAA_bgsubt, results='asis', eval=(PSIavailable && "bgsubt" %in% names(byaa_all))}
cat("## PSIs calculated per amino acids
    
Median protein stability indices per amino acid sequence have been calculated from background subtracted count data.\n")

  htmltools::tagList(
   lapply(unique(targets$experiment), function(x) {
      dplyr::filter(byaa_all[["bgsubt"]], experiment==x) %>%
                      head(n=as.numeric(SHINYREPS_MAX_TABLE_ROWS)) %>%
                      DT::datatable(caption=paste("PSI per amino acid for experiment", x, 
                                                  if(as.numeric(SHINYREPS_MAX_TABLE_ROWS)<nrow(filter(byaa_all[["bgsubt"]], experiment==x)))
                                                  {paste("(max", as.numeric(SHINYREPS_MAX_TABLE_ROWS), "entries)")} else{""})
                                                    )
      })
    )
```



# Used tools and versions for this analysis ##

The following tools were used for data processing:

```{r ToolVersions_paragraph, echo=F, results='asis', error=F, warning=F, message=F}
cat(Toolhelper.ToolVersions(), sep="\n")
```

R session info:

```{r R_sessionInfo, echo=F, results='asis', error=F, warning=F, message=F}
sessionInfo()
```

</div>

