# ChIP-Seq pipeline
Here we provide the tools to perform paired end or single read ChIP-Seq analysis including raw data quality control, optional adapter trimming, read mapping, peak calling, differential binding analysis and functional annotation. As input files you may use either zipped fastq-files (.fastq.gz) or mapped read data (.bam files). In case of paired end reads, corresponding fastq files should be named using *.R1.fastq.gz* and *.R2.fastq.gz* suffixes. The Pipeline performes 2 branches for bam file processing in parallel:
- filtered branch: removes multimapping reads (MapQ filtered) as habitually done in most ChIP-seq studies and (optionally) removes duplicates from BAM file assuming they are unwanted PCR artifacts (duplicate removal is usually recommended for paired-end data but not for single-end data).
- unfiltered branch: alternative workflow using the unfiltered BAM files. This may be preferable when studying some types of repetitive regions (keep ESSENTIAL_DUP="auto" for MACS2 peak calling).


## Pipeline Workflow
All analysis steps are illustrated in the pipeline [flowchart](https://viewer.diagrams.net/?tags=%7B%7D&highlight=0000ff&edit=_blank&layers=1&nav=1&title=NGSpipe2go_ChIPseq_pipeline.html#R7R1Zc9o699cw0z7AeAEDj9lIe2%2B%2FNE3S296%2BZIwtQDfGdmyTpb%2F%2B02q8CCPAG0nTmSaWZVs6Ojq7zunoZ8uXy8D0F%2F%2FzbOB0NMV%2B6ejnHU1T9cEA%2FcItr7Rl1O%2FThnkAbdZp3XALfwPWqLDWFbRBmOoYeZ4TQT%2FdaHmuC6wo1WYGgfec7jbznPRXfXMOcg23lunkW39AO1qwVtUYr298AnC%2BYJ8eaUN6Y2paD%2FPAW7nse67nAnpnafLXsDmGC9P2nhNN%2BkVHPws8L6J%2FLV%2FOgIPByiFGn5tsuBsPOQBuJPPA%2BX%2Bv4fn3fx4ffi%2FOhtczFXyOlG48uOiVwwLYCDTs0guihTf3XNO5WLeekvkC%2FFoFXf23Wvq8%2F9z0Ucv6qS%2BehxrOVdwNRNErW3lzFXmoaREtHXYXvMDoJ35hb8Cu%2Fk3cOX9h3yIXr%2FzCjYLXxEP48t%2FkvfVj5Io%2FF0ZmELGRGPjaWwUWuAYBXIIIBLe%2BaUF3Hj9qJ7qiJ%2Bcg2tA1jALvIcYftL6n%2BSViq0Y%2FSZuerKtvtz9Wf63Of%2F%2B6%2B3w3%2BTd8%2BtXVGeLRDxZ07LMFxKuW%2BARDgUvgoYEGr6hDABwzgk9phDfZvpnH%2FdYIhP5gOCTGp0%2FKX7PF1ShynhY306tPt5P735%2B6fEM%2Fmc6KfaqjGQ6awakNn%2FAXHTh3yQ3jcYVR%2FzQg2yq%2BRH%2FN2W%2Fy2DTItqCBkXfx1gz%2BJrDKMafAOY236JnneAHppE%2FIj3DRZp7LsUPFa26b4YIgu5pCfXw1g46TfKmC%2F8Uv5XcITdBP54FpQ4QImWbLW0KLYdDcMcOQYxMnFwr5bIQWz8OA646VIsx6AkEEXgoxgd0dGXpPY3Sbke2uqqpx23OCEA7Zsi4SNLCv9HuDw9FIiNXqu8ajCwP%2FqwKP4BIxspPQpzxULQeRtJGawqKRkkOgvjLOI9CQ065D0Oe%2F2VX4Zfby8nqz%2BgW1Ifx5p%2FzTHRzM1OpmYWqKgeE9uI2Hlchr8qxGCFW1KU4jHI1xdGuspNdYb3aJxf30pta4aNibmECGwGcbMAFOYQhnDfhGNyRrfII6qH3%2FRcA3zrTOSZJX0NflWEjZn83O5mzx%2Bbp7Cx7RE9fQBw5E1B4N7bRwDEJmt2ECAlBSFnvgDA6DQ5dgIYzMKXoaIT2fOtIxwQy6kEhDCH3wf1vmWM6CmJh7BWCWesMiirCGeoLRXJvMYYQkhV7g%2BD0XbTNtApfTmYe2HPrz6vIWz0Gbe%2Biii9fM8aZ4U5ghUiTQH3yGIfobr3mIl3xiLZAKDB57%2FG4PSSDe02t%2BmJs7ctCYkihRDrhC33Rl2tAgdmuuGYtbvUEcz7RDbIXAq45tDp69cjAGKaZr4w0UhohVQNOhbMMkSmtJNKKE7f0CrFVEhpvY3mjD4NcitZps%2FA8h%2BIh%2BkUbfhAEg8yKz%2B%2BCTWzYIsSogP6vNcn4pwrCSEobzylRfFShTwxJEYbF9QMBCMwBAGoKP%2F0QahI90CTx1KgZx2WPEG2ITWF8IrEIeLg1BNQcxTQAw3iYtiLDvXXuQ4C37WDejAQ%2FTL%2FBmsxBT8swixIM%2BQMNtVEfhf%2F%2BbkmV3k1%2FrVFFUTVaArUl%2B%2FR6C4Ov0P6xHIzKMzQMiI0WeOE69F0wbidEQU8epF9gg6KJmLMdgEHHEJ7RTQVg%2BiO8kjc0nRO7RQR%2F%2Fi3v4pm3H79baJ87Q9Qt70YvAUpO%2BWSS2CNgRFxTzIN8MzyynAsvkoBQLrXYYW3jXt5RcrxjsiV7aZthHU89%2BzTXmGFdk85bP14Xsbd0x%2FyhvgbxhtiLECb9SCc2lT6DmItEg8UIo9RHcmB904TSS3zl8KmQO9JVKhGhNsRBQ1aSurr%2Fflb88rr%2BKNq1Q7dMrednY3JpeOUxK%2FX0nwR7mk%2FBmicWqeuA3wHegZUb7r4m7Wk4xmabjDvZ8384Dvwbmw5npOFuUkMKRJ%2BDto9dhCrzPG4Vjx40i2kw4F%2BEvSR2iWL7TEoIHEwgTsh4WhxHInRPmxoiwAHfKnRoOmOHPeKjXzCHGe%2Bwh6KREuTJ0lbTdXtVFordA9lYHuwtM6DIhM%2B0gKGujP4KyvKDM1bujEZR9ocy2RMOHLhXLFH8tG9P2Ltkt%2BJ6RuIeQPuqyHXRCBDOXWju2WViv18aHnJFE2rjqZ9sWuV57TZWSAnxzJBbnTztrA%2BgOljRRWzM6QWyk6j2ZQbjRwLmpW15PYLpRwWJtgFNuDatZsYwk4QfeE7TBNnOdSJxITu40tv9doQdTQEoBRPBd%2BlgXu4XhDFrprysfPD%2FqfZQdhgiYW5mlVswsd2GLzwsYARyXg599Dkg8UvmsUh8N5VjlePcIif05pchdVqutj3%2Bwtca%2BfmoV%2BxUY%2B4RRdqqaW4g3J8PUFVOXCqPZX1DSpV3iRtmSEnn0JAjM10QHH6NquBl3h2rar6AyU%2FVEsr8%2B1jKYTUcgiefbxLhYSJogZv%2FtjLAcGQnqcYWIe4QBYXkIoTwn1knNZwJT4uDC80CCw0OHOWxKk0hihxnuih1F6Maj1Vugz%2BL7WxkXxaKNjEtEbTPMjBkYv5Crc93YwMJIMBl7sVK6IjhQhilkGee5m2oI6LKxu9dKhrv9%2FhGd3yk%2Ff%2F01ubkIlqp9d%2Fn6uauOC2kocKbec%2BPkszpy1ZckV4PS44H3Ilf9jBu0z9T4PclPDkUmN4Z1%2BfN8cQO%2BPFqzf1%2Fvvp797jZqKVBSaDLci8vWGhImi1H8KEf9MWHCRVb1Rle5lvMJ1a6yEKylSzkHLbJ%2B%2BCKXeiwlzQOG7bcDSu9uftCkaX5hpPmFxgLGN4m32f66Ui5%2FCW4vrhfqdGXMvj%2Bew%2FNT7%2F4EdhulPGVocbvzlwrFF%2BHGLx0bZcmQcMH7pVIhdlCj5TiQWvFmBI52IUGx3vEHCSpCgmGrkEA9%2FLjJHyzYAwsGNWGBtH1r6j1HEGjSBi7ekghZuwGmjb0upu9jQ%2BcWX9Wm6HPa9CEk0RDYT%2BPTwxE4GPrjpudrMKStHXsMVDtY0%2FqJxa3SmlZyGAUzrcamWYOhd0Pms0IztuAs1XYvJwnhZ75OZQ%2BX%2B9dVREO%2BcGIG1GG9eRI7R3m0irZUcz52itt4m8XTgPSUBT2rsB63jCP08CCg5MFiftB4Yi6hg%2FHvE3CeAH5rpw636LA%2FSqthhpZHfb49DkR9STwftgzPlRyqByBcOVgZbiOyf0BIhuYG7I%2BdOMCOoXu9cRLncDYDwToegg7FgWFU%2B1Cu6bdN1%2BVpGFq2uyvYyX1DkAZCtJNHWlU7edT6nex485ZuYzSyLudSluc4SFogJ%2BcoxwKmtejwuJ93gMyGLonMZbClv58u7%2F2zu%2B%2FR1RIY18vVVXR%2Fdfymw23G6%2F2VQZ4FZKsyOG7MR1E47tbQp7xEHQAfoU1LSVQ6dLXq2FOcTg6rpzFIJuECuq%2F37Nz8PW1HuuM97sn0WNrYu1na%2BeHn7ucDUTdEabZ6yj3SB91CarvjgyDs3YjnLu64EQiNhdfiRUJUKd4NeM8E3rKTPnSelHClzyA1MyHEteHslTCzCHNxaxVGdD48aHedACBC%2BBUuEInATRDHVbkAYJ4iPbE3r7IOVUnZQFVKyHY2ETm9%2BoeHh1bo7a5GXqjMttw%2F8sBOjVkP2xbY%2BdlFPdkmlrR9Y%2FjjFD34Qbx0%2BFlOfQ1zibHUnYZ0XytTc8kIsUjTork%2BRB9l0KglPJROBc9D3qjdrytENLGRM%2FkEAfnpZPMJoju6oY91u5PIElhRRKmm94WJJWs0i4sj8w938R8XoW1%2F6lvu8ztWAq73jVYScH9hukhQJIZUWQr%2B7QzLqejb0AopWaZkm%2F6PqTZ0berEtBZI%2FLQQZsAwintniHi4iYjX6pdMAGIHMj74Q8YRrvI4iIZouDg24y2GaQ3rjBZlFEuC5JYelVNKtKjBNCjZaNFB%2BvBUZST38ubi5E6a2F64iMwuloCM3ESI9xrCMPfOQitFUySVTHQHYmrUT0xTpHF3yloCMc3GiRj8WGibjllpx3DMqrwI%2B5Lim3m04lYCylOlNy2z6hkCmq4dUhlBREr%2BGbYOknIskmTRCgBJTaS4XrBEZCAkjswpnD9DLHdGSOp8qPOMKJqDxeawA8Ub1k%2FxKpD%2FBkazJ0N%2FuPrD8%2Bnj3a8f1%2Fc37t%2B%2B%2Fvnpvlnf6rEdCOPW7ySlEkK1sWz%2FQhM5T8z6lkzk9Yr4gnU%2FJhG%2FPy42izcl4mMrwn0iMkyaqyFKbwYwJFnsqbPuCYY4G8JvFmJWr00kPY04vdJ2xjZqWJTPiewG%2BelQJ2NSlCc%2FeeH%2FgvxUI%2BL3Db1Rfim2eWtNMsw%2FNu%2BCI2%2FHavPus1P5rbN5Ty1pinx9doPedupFkQNcYOGI3zMPzGbQgsQkUy89nu6QhYZiz7u3TY%2FaSGvffJr%2Fo6O1PEPO0dLaUTsDRKCPVgm4c5ZsX4bk%2Bo6H30pz0fOH6yW061HL09uB0qmd3lZAMPuDZs8pCu3P%2BvHVbmzSmiN9kHpUNi07zJxzeKqud27Okc7v01Jzjrpbfp%2B6zDlL0wrlD%2FcnChCQIJhaORcZ6g5Kwts4XK%2BPxo0yrUKvXmuOAuVjr7ibrI0ngaZw%2FiPly2PHFJlBNEyduW3vGfsSsD17OEHob6vq4OJ0uLIvvaufPlh2v92Mo7F1aTRbSvvYqm%2BMpLlyU9JY4bD3Il%2BVl%2Be4ASQPgIVDAC6B66UqlcmFRTV6wvH064%2B7zxeEP6Nd8UKiF%2BapeWQPCPK7Ck5HjU8JIoa%2FohCo%2B%2BwZmQD0cE2IZPoDtDusBxzEQQtgFR%2Bda7qsBjtT4ngWGT05CofnsK2iyF51LlpWtiItPBlDQcTZeJxnJ6paGa0ZCGhNBnDVFqmIqd37LVIhTFHQ6EmYtOElZu47GV7UGjn9WBAxLbR0DOrKYCe3%2Fcab04C1vmRtzOsTR%2B8JKhUrNe%2BtWCwFSTpZGfr1ASG5DWj9dHYAPpvCgJ8of3BhFHxcX8%2BgS3IwxYmraIkOk6CbqIIHjWdZh7AjQAM%2F7CUWRabWZlVVKHd8%2FzsZWGen2pellrKsSPgZG2qKk4pP%2BovqW1Z20j8v6zSmTR9BEntVlT4K2iomG4%2B7EXsgL69cnBnIfLZNrFu20R5IqjB11mk2P%2FyGvk8yEjasUyJdlx129dwnrDsSXZgaK30TUuYKmKzTwtQxtEC12rtRewTGvfnvdo9T22OcR20AztSZlK0yOa6IkvHqIC12bOA02qmERi0iZOucJtIWsqPG3qz7QuNVwLa5LwaVCVyH%2By%2FaV0io1kiDsSopg%2FF8iO3J1U8y1N2w7HeyPn0GEMzjqfWYcHe03x9s79ntrNPI4VHX5uznyfbwVOR9%2FuPkChyvz3%2FYb9bn%2F0n5a7a4GkXO0%2BJmevXpdnL%2F%2BxM%2FHyfgjDnkwtRXyF2yNqW%2BiIl8AXMsUq6Rl75u29n%2FxDpjhptepHSUNst4kQzpZk0cGxiL1k%2Bz%2FGgJbZvQTBFWpOloGfyFx%2B7ySgt59tIXyUfa7nghhwT5ittTvLc7idzKHU1%2FOenQxOoOURBIHkbPx9qD6XzMrdZ%2Bu7IM6GZqduqCXC99AXT7VUH3ELe0kRU%2BE1ZYDs11I4GrSC9JZalLJAxdL07jQuaekuUuBroykGs4TiHXSBDYIkSuEiRDIXLV6YcsxO7364YUguUIHVTMPj6hnpaC3fjeXFGFwHi3npbWDqxBF1AJHGY83E980fubaeNBHKbJSkSQOgEmXsaG1soouXdgGcsip7T4Y1SEnJuL68Tm9c8RgqaVjBlfW97Jcq5CjFEKw7U4tfyb1T6HaeumKkthKlM%2F89b5c3oal1imAI10tBHESP4sUaQGCcx4y0uWlmLVfn7JNOGSleBREcbTv%2Fl803vanoXAkjU91%2BX%2B32Z5bkG0%2FHXSRuHO4HwV8KjnNnrI0hVxkNbiCsYmaq66eA4FHr6B%2B%2FXo5TpGPOvoE3aKxZqJYFINVshJgrw2OPoBMJdIiQg3AlHQowXAyghFaeiJohGqhii3QqYdRKTY00bQZu62EayZBhvMKBV79oIHx2MVs11zTkTOIviL2g5pCADlU%2BEO3%2F1Q3gARIMyVQwqpsBEwWIbOKkgZfnCuflrjyUuK40q4%2FSOkUAt4ARYSHYLUOuNTM05qAaXmhaOHaRSw6O76bE5cWzne%2FPGDtChZ6tjOB%2BKLfQAA4%2BQaMCESfqA7Dz%2F2tgwvo00WCmrFrtMWhLdmQoUG3PG1LVhIHWg7Cz4yXlIhFDWBmlmVlb1IhHzHVvaJsORYcZrvo0zAUadmIp1NtfS8K3ul38jm%2Bx6yNDsbc0hl84MbgwwWVpN%2Bg5UqJuK0fMiO5fmvZDFCUiKcFGVG7KOTOATDGrkVVOtUHruzhCF2sZgu8FYhEYQTU%2BvRfj1x%2Fo7NGHd0oTwDNc2lhkqzRcDExDDPlBolhkpvJ3KILuK0eVkSqTZHIkWHN4Tg51E8LclNxQtIvn1s0JpGh8KA07agQ8tSlZWIDhn5SW8YG4oErfqRoXDYUvrMKnBeT3FGIzzBbbx2jTH0iqeTP%2B%2BOd4S9PHPuDwe9QVr%2FGCs90SHJPH%2FWh1VtuC0ViI5SNakqYr%2BocsexpgYcDFO1iKpODShZY7BcoeDwGoMtzlR6EFL2S0%2B6fOgpEhvOZrjMqbQ6ih%2FAGatwaqFOskTqOiFCDfrn2jXAJ6ALlc3Nq1Wvsrl7HY%2Biin5VFvUY6IaISQq12N3xUkaL1eZQHX%2F9eTsZAuv04u%2FX%2Bewz5DvsT2RBhhwJhYy2JRbYRo1aG%2FvcrNeTBDKhLuE9J3K96EVwkmNjvyJP6HuN1o49Z%2FToI4bYBjaGPY4ejzUhAMY4Qzy0NAYFdwC9eY9O%2Fe%2BvT%2BGPOzT%2BD39%2F7f642%2BquKzGeOn79JcJ37D1kaQvxHmXpACktCDOIccC4OjuFUxfqRMXFx1ubgUdLM05tPMqxTaGLclij8bfZalgtFuOF0JJNe1e6anmYSeFw2ah1JoU6pSXp8vDjVloUNGO3YgO6MjrEoiCtWSI6PPWeIwi0lQsfV0BawwzAEpFz9PblyolgF6dAoaw5QApeMk4pfgc%2BfRz1MBsPTNdafBR%2Fp1a1NDt7oXa6GR1r1k4b0jU1vqXa5C9VG1U1j41lypaG5yVd2sIzD5eL3jXPlF32lvJMvb9bveUDeWYOJ%2F%2BbXYVfZi8vrzerX1Abwp93yj%2Btqly7b%2B6mLSiYKaeYxUeBE6EGytQWK1gsqtwsz1e%2BfIBYLC%2FZK98hx%2BHsoxCWUFf0f7BEwxZnfdq8rPWKR%2Bhl0OJEtCFRSVebLWco9BuWm37ucLdhAxVhy2OpY1mW2s7CrWq6EGtrCreemstvZ9K0VHSKmKSDXye7rJE2oo8%2Biitlb8ag4wuizRRuVQ2tfZSuhErXVZI6rS6NIUX99qd1nBYcK7EzxqmonNYQu%2F9hi9kO5C5cLZdmgHFNU76ddeJsvrVROGLi24XGUcAeHZHLinPGqFkiJ%2FbDbU7sTz30%2BdSCARnpbllvUp7nDFFNQJ3g%2BWnsxE%2FI2RPyI6RPyQQ4RpZepaNQs8K7gf91NiQDmQemDcE6Uoc1JzWDuWOGYVZLUMpBn3HWHsGvtxX70PWqck8PtqWy2yEPJXlq5WRbnPUpZfMZ67ekpoMw04vveFEyCYST82wnXgbIeesogEtW8AB7j58geCa4aK5fJERfiVevXRb4dTCM0omGit6Aqe66Ruz3s1tMlXlhSAWh8DNNSxQXkyUnwVgtWULH2RgsPClSxHDPuUAXfYkcQKbvzU9l51ealgXCkELGgdPADOgq4noIL3RJp%2Fi%2F67Mb%2FKcXRQ5wgUWqUHtgNoMWTCclLfoYlhrcAFqLJXmGowiDbHIon6%2FJ9ND%2BnOOKUFJvtwIvDMm4Aip60MKR7Bu5RfEXphuRslS4qrYsMuDS28SYwx5SeIQOPUP%2BD3DxR21oIvK0XM%2BNFphgeO2YvuzqpEpgpqDFS4HHOMVGYCOECOB0xR66m3SSoTSOZ0HpjcOLb63DcrZEme6Eg6gxT1%2B28KLD80QxieOALFEpc1OS25TBVPrqVqYSJ6FOMRVeueMQpiI0gTdqAW%2Bx603sL5DVoRpzvYmH3bITiG0uHCa77uJ%2BpR813M%2F3Ns4YCreczs%2F217R64lXOVpFpm34krTyT3kwY46JkbYqzxUYr1Jw3o00NijNRwdiLlfSW2%2BQvSXG5CrTtsUBbqlHZLtydLS4%2BRdC6rdWn4sEpmYJ6BXLhW8m3m6tEJciOVFYlKuyi9bwoSW7RPl%2F8z7MxJb34Pw%3D%3D). Specify the desired analysis details for your data in the *essential.vars.groovy* file (see below) and run the pipeline *chipseq.pipeline.groovy* as described [here](https://gitlab.rlp.net/imbforge/NGSpipe2go/-/blob/master/README.md). A markdown file *ChIPreport.Rmd* will be generated in the output reports folder after running the pipeline. Subsequently, the *ChIPreport.Rmd* file can be converted to a final html report using the *knitr* R-package.


### The pipelines includes
- raw data quality control with FastQC, BamQC and MultiQC.
- adapter trimming with Cutadapt (optional)
- mapping reads or read pairs to the reference genome using bowtie2.
- filter out multimapping reads from bowtie2 output with samtools (performed in parallel pipeline branch).
- identify and remove duplicate reads with Picard MarkDuplicates (performed in parallel pipeline branch). 
- generation of bigWig tracks for visualisation of alignment with deeptools bamCoverage. For single end design, reads are extended to the average fragment size.
- characterization of insert size using Picard CollectInsertSizeMetrics (for paired end libraries only).
- characterize library complexity by PCR Bottleneck Coefficient using the GenomicAlignments R-package (for single read libraries only). 
- characterize phantom peaks by cross correlation analysis using the spp R-package (for single read libraries only).
- peak calling of IP samples vs. corresponding input controls using MACS2.
- peak annotation using the ChIPseeker R-package (optional).
- differential binding analysis using the diffbind R-package (optional). For this, contrasts of interest must be given in *NGSpipe2go/pipelines/ChIPseq/contrasts_diffbind.txt* (see below). 


### Pipeline-specific parameter settings
- targets.txt: tab-separated txt-file giving information about the analysed samples. The following columns are required
  - IP: full IP sample name without any filetype suffixes like ".R1.fastq.gz" or ".bam" (sample name itself must not contain dots).
  - IPname: (shortened) IP sample name to be used in plots and tables.
  - INPUT: full input control sample name without any filetype suffixes like ".R1.fastq.gz" or ".bam" (sample name itself must not contain dots).
  - INPUTname: (shortened) input sample name to be used in plots and tables. 
  - group: variable for sample grouping (Condition).
  - Replicate: number of replicate (needed if differential binding analysis is selected).
  - PeakCaller: name of peak caller (in this pipeline it is macs; needed if differential binding analysis is selected).
  - optional further columns which may be used in multifactor model design: 'Tissue', 'Factor', 'Treatment'. 

- essential.vars.groovy: essential parameter describing the experiment including
  - ESSENTIAL_PROJECT: your project folder name.
  - ESSENTIAL_THREADS: number of threads for parallel tasks.
  - ESSENTIAL_SAMPLE_PREFIX: sample name prefix to be trimmed in the results.
  - ESSENTIAL_BOWTIE_REF: full path to bowtie2 indexed reference genome.
  - ESSENTIAL_BOWTIE_GENOME: full path to the reference genome FASTA file.
  - ESSENTIAL_PAIRED: either paired end ("yes") or single read ("no") design.
  - ESSENTIAL_READLEN: read length of library.
  - ESSENTIAL_FRAGLEN: mean length of library inserts.
  - ESSENTIAL_BSGENOME: Bioconductor genome sequence annotation package.
  - ESSENTIAL_TXDB: Bioconductor transcript-related annotation package.
  - ESSENTIAL_ANNODB: Bioconductor genome annotation package.
  - ESSENTIAL_DB: UCSC assembly version for GREAT analysis (only for UCSC hg19, hg38, mm9 and mm10)
  - ESSENTIAL_BLACKLIST: path to bed file with problematic 'blacklist regions' to be excluded from analysis (optional).
  - ESSENTIAL_ADAPTER_SEQUENCE: adapter sequence to trim with Cutadapt (optional)
  - ESSENTIAL_NEXTSEQTRIM: most Illumina instruments use a two-color chemistry like the NextSeq (exceptions: MiSeq, HiSeq). This option accounts for terminal G bases incorporated by faulty dark cycles during base quality trimming with cutadapt.
  - ESSENTIAL_MACS2_BROAD: use broad setting for broad peak calling in MACS2 (default false).
  - ESSENTIAL_DEDUPLICATION: remove duplicated reads in filtered branch (default false is strongly recommended for single end data). 
  - ESSENTIAL_DUP: how MACS2 deals with duplicated reads or fragments.
  - ESSENTIAL_MACS2_GSIZE: mapable genome size for MACS2
  - ESSENTIAL_DIFFBIND_VERSION: DiffBind version 2 or 3 to use.
  - ESSENTIAL_DIFFBIND_LIBRARY: DiffBind method to calculate library sizes. One of "full", "RiP", "background" and "default" ("default" refers to method "full", see [DiffBind documentation](http://bioconductor.org/packages/release/bioc/vignettes/DiffBind/inst/doc/DiffBind.pdf) for explanation). For DiffBind version 2, any other method than RiP refers to TRUE, i.e. use total number of reads instead of reads overlapping peaks.
  - ESSENTIAL_DIFFBIND_NORM: DiffBind method to calculate normalization factors. One of "lib", "RLE", "TMM", "native" and "default" ("default" refers to method "lib", see [DiffBind documentation](http://bioconductor.org/packages/release/bioc/vignettes/DiffBind/inst/doc/DiffBind.pdf) for explanation). Not applicable for DiffBind version2.
  
- additional (more specialized) parameter can be given in the header files of the individual pipeline modules (see module header files linked in the flowchart for default parameter).

If differential binding analysis is selected it is required additionally:

- contrasts_diffbind.txt: tab-separated txt-file giving intended group comparisons for differential binding analysis (1 contrast per line). 
  - contrast.name: name of contrast to be used in report.
  - contrast: definition of contrast using the format '(group1-group2)' with 'group1' and 'group2' being groups defined in the 'group' column of the targets.txt file (pairwise comparisons only).
  - mmatrix: design formula to be applied for this contrast. If no further factors to be included use '~group'. A multifactor formula must be composed from the following allowable factors: Tissue, Factor, Treatment, Replicate, Caller and group (aka Condition). For diffbind2 multifactor design is not supported and this column is ignored (group column is used automatically).

This pipeline supports 2 different DiffBind versions (v2 and v3) for differential binding analysis. Beginning with version 3, DiffBind has included new functionality and modified default settings. The major change in version 3.0 is in how the data are modeled. In previous versions, a separate model was derived for each contrast, including data only for those samples present in the contrast. Model design options were implicit and limited to either a single factor, or a subset of two-factor "blocked" designs. Starting in version 3.0, the default mode is to include all the data in a single model, allowing for any allowable design formula and any set of allowable contrasts. 

**IMPORTANT NOTE:** Mind that for DiffBind 3 the contrasts are processed isolated from each other. This means that a consensus peakset is generated separately for each contrast and does not incorporate peaks called in other pulldowns unrelated to the contrast. A pipeline update that allows processing together pulldowns from different contrasts is currently under construction.

There are also more normalization options available in DiffBind v3 and a new option to generate greylists as described in the [GreyListChIP](https://bioconductor.org/packages/release/bioc/vignettes/GreyListChIP/inst/doc/GreyList-demo.pdf) vignette (see diffbind3.header for default settings). The purpose of greylists is to identify regions with anomalous signal in the input or control sample, so that reads in those regions may be removed prior to analysis. 


## Programs required
- Bedtools
- Bowtie2
- Cutadapt
- deepTools
- encodeChIPqc (provided by another project from imbforge)
- FastQC
- MACS2
- MultiQC
- Picard
- R with packages ChIPSeeker, DiffBind, GenomicAlignments, spp and genome annotation packages
- Samtools
- UCSC utilities
